internal/planner/type_join.go

```diff
@@ -568,54 +666,90 @@ func (join *invertibleTypeJoin) Next() (bool, error) {
 		}
 	}
 
-	hasFirstValue, err := join.dir.firstNode.Next()
+	firstSide := join.getFirstSide()
+	hasFirstValue, err := firstSide.plan.Next()
 
 	if err != nil || !hasFirstValue {
 		return false, err
 	}
 
-	firstDoc := join.dir.firstNode.Value()
-
-	if join.isSecondary {
-		secondDocs, err := fetchDocsWithFieldValue(
-			join.dir.secondNode,
-			// As the join is from the secondary field, we know that [join.dir.secondaryField] must have a value
-			// otherwise the user would not have been able to request it.
-			join.dir.secondaryField.Value(),
-			firstDoc.GetID(),
-		)
+	if firstSide.isPrimary() {
+		return join.nextJoinedSecondaryDoc()
+	} else {
+		primaryDocs, secondaryDoc, err := join.fetchPrimaryDocsReferencingSecondaryDoc()
 		if err != nil {
 			return false, err
 		}
-		if join.dir.secondNode == join.root {
-			if len(secondDocs) == 0 {
-				return false, nil
-			}
-			for i := range secondDocs {
-				secondDocs[i].Fields[join.subSelect.Index] = join.subType.Value()
-			}
-			join.docsToYield = append(join.docsToYield, secondDocs...)
-			return true, nil
+		if join.parentSide.isPrimary() {
+			join.docsToYield = append(join.docsToYield, primaryDocs...)
 		} else {
-			secondResult, secondIDResult := join.processSecondResult(secondDocs)
-			join.dir.firstNode.Value().Fields[join.subSelect.Index] = secondResult
-			if join.secondaryFieldIndex.HasValue() {
-				join.dir.firstNode.Value().Fields[join.secondaryFieldIndex.Value()] = secondIDResult
-			}
+			join.docsToYield = append(join.docsToYield, secondaryDoc)
 		}
-	} else {
-		hasDoc, err := fetchPrimaryDoc(join.dir.secondNode, join.dir.firstNode, join.dir.primaryField)
-		if err != nil {
-			return false, err
+	}
+
+	return true, nil
+}
+
+func (join *invertibleTypeJoin) nextJoinedSecondaryDoc() (bool, error) {
+	firstSide := join.getFirstSide()
+	secondSide := join.getSecondSide()
+
+	secondaryDocID := getForeignKey(firstSide.plan, firstSide.relFieldDef.Name)
+	// TODO: add some tests with filter on nil relation
+	if secondaryDocID == "" {
+		if firstSide.isParent {
+			join.docsToYield = append(join.docsToYield, firstSide.plan.Value())
+			return true, nil
 		}
+		return join.Next()
+	}
 
-		if hasDoc {
-			join.root.Value().Fields[join.subSelect.Index] = join.subType.Value()
+	if !firstSide.isParent {
+		for i := range join.encounteredDocIDs {
+			if join.encounteredDocIDs[i] == secondaryDocID {
+				return join.Next()
+			}
 		}
+		join.encounteredDocIDs = append(join.encounteredDocIDs, secondaryDocID)
 	}
 
-	join.docsToYield = append(join.docsToYield, join.root.Value())
+	// check if there can ever be false (a.k.a. hasDoc = false)
```

Comment:
question: This comment looks like a temp dev-todo, is it done, should it be removed?

Reply:
removed